packages/sandcastle/templates/Sandcastle.ts

```diff
@@ -0,0 +1,237 @@
+let defaultAction: (() => void) | undefined;
+let bucket = window.location.href;
+const pos = bucket.lastIndexOf("/");
+if (pos > 0 && pos < bucket.length - 1) {
+  bucket = bucket.substring(pos + 1);
+}
+
+type SelectOption = {
+  text: string;
+  value: string;
+  onselect: () => void;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any -- we don't care what the keys are
+const registered = new Map<any, number>();
+
+/**
+ * Helpers for constructing UI inside a Sandcastle and interacting with the code editor
+ */
+const Sandcastle = {
+  /**
+   * Called on first load and every time the options set up by other helpers are changed.
+   * No-op function by default, override with custom reset logic when needed
+   */
+  reset() {},
+
+  /**
+   * Create a "bookmark" of sorts in the code that will be highlighted when run
+   *
+   * @param key
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- we don't care what the keys are
+  declare(key: any) {
+    /*eslint-disable no-empty*/
+    try {
+      //Browsers such as IE don't have a stack property until you actually throw the error.
+      let stack = "";
+      try {
+        throw new Error();
+      } catch (error) {
+        if (error instanceof Error && error.stack !== undefined) {
+          stack = error.stack.toString();
+        }
+      }
+      let needle = `${bucket}:`; // Firefox
+      let pos = stack.indexOf(needle);
+      if (pos < 0) {
+        needle = " (<anonymous>:"; // Chrome
+        pos = stack.indexOf(needle);
+      }
+      if (pos < 0) {
+        needle = " (Unknown script code:"; // IE 11
+        pos = stack.indexOf(needle);
+      }
+      if (pos >= 0) {
+        pos += needle.length;
+        const lineNumber = parseInt(stack.substring(pos), 10);
+        registered.set(key, lineNumber);
+        console.log("registered", key, "to", lineNumber);
+      }
+    } catch {}
+  },
+
+  /**
+   * Highlight the given "bookmark" in the code
+   *
+   * @param key
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- we don't care what the keys are
+  highlight(key: any) {
+    console.log("highlight", key);
+    console.log("registered", registered);
+    if (key !== undefined) {
+      const lineNumber = registered.get(key) ?? registered.get(key.primitive);
+      if (lineNumber !== undefined) {
+        window.parent.postMessage({ highlight: lineNumber }, "*");
+        return;
+      }
+    }
+
+    window.parent.postMessage({ highlight: 0 }, "*");
+  },
+
+  /**
+   * Signals to the page that Sandcastle has finished loading and calls the default
+   * action if it has been set.
+   * This is called automatically as part of the loading process, you should NOT need
```

Comment:
Do "private" members like this one belong in another file? I'm hesitant to expose anything to the user that they are not supposed to use.

Reply:
It needs to be exposed to the sandcastle somehow, this is the easiest place for now. Maybe we can find a way to lock it down a bit more but I think it's fine for now. This PR was targeted at just pulling the code out into the larger build system, not necessarily changing and optimizing this API.